### DISCUSSION TITLE
Domino REST development

### DISCUSSION THREAD

**Discussion by tdari** [Tauffer-Consulting/domino#226] [Q&A] [answered]
Sorry if my question is out of the project context, but what is your preferred way to develop the rest api. I tried to connect to the container after started all containers using dev version of the compose file, but I could not manage to debug. When I try to run the rest service locally and all other services using the compose file, my local rest server can not find airflow web server host because of the network settings in the compose file. I would like to hear how do you develop, thanks.

**Comment by vinicvaz** [comment, marked as answer]
Hey @tdari 
You have few options I would recommend.

**1. Configuring python debugger within containers:** This is a bit tricky and may require some time to configure but after configured it should work well without any additional changes. I use vscode and its debugger, so I had to configure it to make it work.

**2. If you are developing a domino only feature:** If you don't need to interact with airflow and other services you can always run your rest locally only and use the default debug mode. For that I'd recommend you to start just the domino database service and develop within these two services. This is the easiest but limited option.

**3. Running all services in docker except rest:** Like you said, the default configuration for the **airflow client** in the rest side and **rest client** in airflow side are the docker network services ips. If you look to `rest/core/settings.py` you can see a variable called `AIRFLOW_WEBSERVER_HOST`, you can change it, passing to your rest environment the value `http://localhost:8080`, it should be enough to make the rest find the airflow, since when running with compose all services are open on localhost.  I think it is easier and faster than setting up the debugger within containers but it requires you to pass this env var every time you run your rest.

Usually when I'm developing REST, most of the changes don't need interacting with airflow then I go on the option 2 because it is faster, easier and lighter. Anyway, all of them should work in their specific cases.

**Reply by tdari** [reply]
Thanks, @vinicvaz I'll try your advices.

**Reply by tdari** [reply]
I can work with the third option but I had to export AIRFLOW_ADMIN_USERNAME and AIRFLOW_ADMIN_PASSWORD. Also, I changed domino postgres port to 5433 because of the port mapping setting. In case anyone reads this in the future.

**Reply by tdari** [reply]
I have a problem though. When I run the default image filter workflow with this configuration, it causes a name resolution error in domino worker. I could not see any option to set the base URL. Am I missing something?

```
def execute(self, context: Context) -> Optional[str]:
        """
        Code from here onward is executed by the Worker and not by the Scheduler.
        """
        self.domino_client = DominoBackendRestClient(base_url="http://domino-rest:8000/")
        # env var format = {"name": "value"}
        self._prepare_execute_environment(context=context)
        return super().execute(context=context)
```

```
Traceback (most recent call last):
  File "/home/***/.local/lib/python3.9/site-packages/urllib3/connection.py", line 174, in _new_conn
    conn = connection.create_connection(
  File "/home/***/.local/lib/python3.9/site-packages/urllib3/util/connection.py", line 72, in create_connection
    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
  File "/usr/local/lib/python3.9/socket.py", line 954, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -3] Temporary failure in name resolution
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/***/.local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 714, in urlopen
    httplib_response = self._make_request(
  File "/home/***/.local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 415, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/home/***/.local/lib/python3.9/site-packages/urllib3/connection.py", line 244, in request
    super(HTTPConnection, self).request(method, url, body=body, headers=headers)
  File "/usr/local/lib/python3.9/http/client.py", line 1285, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/local/lib/python3.9/http/client.py", line 1331, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.9/http/client.py", line 1280, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.9/http/client.py", line 1040, in _send_output
    self.send(msg)
  File "/usr/local/lib/python3.9/http/client.py", line 980, in send
    self.connect()
  File "/home/***/.local/lib/python3.9/site-packages/urllib3/connection.py", line 205, in connect
    conn = self._new_conn()
  File "/home/***/.local/lib/python3.9/site-packages/urllib3/connection.py", line 186, in _new_conn
    raise NewConnectionError(
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7fcb202e2130>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/***/.local/lib/python3.9/site-packages/requests/adapters.py", line 486, in send
    resp = conn.urlopen(
  File "/home/***/.local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 798, in urlopen
    retries = retries.increment(
  File "/home/***/.local/lib/python3.9/site-packages/urllib3/util/retry.py", line 592, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='domino-rest', port=8000): Max retries exceeded with url: /pieces-repositories/worker?workspace_id=1&url=https%3A%2F%2Fgithub.com%2FTauffer-Consulting%2Fdefault_domino_pieces&version=0.8.0&page=0&page_size=1 (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fcb202e2130>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution'))
```

**Reply by vinicvaz** [reply]
@tdari really, the option 3 is not the best option for developing with full domino. The problem is that the custom domino docker operator `src/domino/custom_operators/docker_operator.py` is trying to access the domino api using the internal container ip `http://domino-rest:8000/` but your rest is running on localhost now. This url is hardcoded in the `docker_operator.py` file and cannot be dynamically configured. If you have hot reload active you can edit it but we can think a way to make it configurable based on environment.

**Reply by tdari** [reply]
Thanks again @vinicvaz I can test airflow separately not a big problem.
